### ISSUE TITLE
What to do with exception type name in error message?

### ISSUE THREAD

**Issue by sungam3r** [graphql-dotnet/graphql-dotnet#3268]
`ExposeExceptionStackTrace = true`:
```
GraphQL.ExecutionError: Exception of type 'System.ArgumentNullException' was thrown.
 ---> System.ArgumentNullException: Exception of type 'System.ArgumentNullException' was thrown.
 ---> System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values.
   --- End of inner exception stack trace ---
   at GraphQL.Tests.Errors.ErrorInfoProviderTests.exposeExceptions_with_real_stack_trace() in C:\_GITHUB_\graphql-dotnet\src\GraphQL.Tests\Errors\ErrorInfoProviderTests.cs:line 166
   --- End of inner exception stack trace ---
   at GraphQL.Tests.Errors.ErrorInfoProviderTests.exposeExceptions_with_real_stack_trace() in C:\_GITHUB_\graphql-dotnet\src\GraphQL.Tests\Errors\ErrorInfoProviderTests.cs:line 170
```

`ExposeExceptionStackTrace = false`:
```
   Exception of type 'System.ArgumentNullException' was thrown.
```

We use `Exception.ToString()` when `ExposeExceptionStackTrace = true` and it appends not only stack trace in message but also prepend it with exception type like `GraphQL.ExecutionError: ` above. It worries me from time to time. In addition, I have been doubting for a long time that the stack trace should be displayed together with the error message, and not as separate field in extensions. 

For example error payload may look like in case of `ExposeExceptionStackTrace = true`
```json
{
    "errors": [
        {
            "message": "Exception of type 'System.ArgumentNullException' was thrown.",
            "extensions": {
                "stacktrace": "---> System.ArgumentNullException: Exception of type 'System.ArgumentNullException' was thrown.
 ---> System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values.
   --- End of inner exception stack trace ---
   at GraphQL.Tests.Errors.ErrorInfoProviderTests.exposeExceptions_with_real_stack_trace() in C:\_GITHUB_\graphql-dotnet\src\GraphQL.Tests\Errors\ErrorInfoProviderTests.cs:line 166
   --- End of inner exception stack trace ---
   at GraphQL.Tests.Errors.ErrorInfoProviderTests.exposeExceptions_with_real_stack_trace() in C:\_GITHUB_\graphql-dotnet\src\GraphQL.Tests\Errors\ErrorInfoProviderTests.cs:line 170"
                "code": ""ARGUMENT_NULL"",
                "codes": [
                    ""ARGUMENT_NULL""
                ]
            }
        }
    ]
}
```

instead of 

```json
{
    "errors": [
        {
            "message": "GraphQL.ExecutionError: Exception of type 'System.ArgumentNullException' was thrown.
 ---> System.ArgumentNullException: Exception of type 'System.ArgumentNullException' was thrown.
 ---> System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values.
   --- End of inner exception stack trace ---
   at GraphQL.Tests.Errors.ErrorInfoProviderTests.exposeExceptions_with_real_stack_trace() in C:\_GITHUB_\graphql-dotnet\src\GraphQL.Tests\Errors\ErrorInfoProviderTests.cs:line 166
   --- End of inner exception stack trace ---
   at GraphQL.Tests.Errors.ErrorInfoProviderTests.exposeExceptions_with_real_stack_trace() in C:\_GITHUB_\graphql-dotnet\src\GraphQL.Tests\Errors\ErrorInfoProviderTests.cs:line 170",
            "extensions": {
                "code": ""ARGUMENT_NULL"",
                "codes": [
                    ""ARGUMENT_NULL""
                ]
            }
        }
    ]
}
```

In other words I think it is worth to keep error messages "clean" without additional techical info - without stacktrace and exception type name.

**Comment by Shane32** [comment]
I would generally agree but I think we should maintain backwards compatibility in some manner - not necessarily source compatibility but some option or easy code snippet to retain old behavior. 

I do like ToString in that it gathers all of the data - message, type name, etc into one string. In some of my applications I expand on this and recursively examine and return inner exceptions’ ToString data. Not sure exactly what I would do here without looking at it in more detail.

**Comment by sungam3r** [comment]
Full backward compatible API:

```c#
public class ErrorInfoProviderOptions
{
        public bool ExposeExceptionStackTrace { get; set; }

        public ExposeExceptionStackTraceMode ExposeExceptionStackTraceMode { get; set; }
}

 public enum ExposeExceptionStackTraceMode
 {
        Message,
        Extensions
 }
```

**Comment by Shane32** [comment]
Sure. Looks good to me
